sql/src/main/java/io/crate/operation/projectors/BlockingSortingQueuedRowDownstream.java

```diff
@@ -0,0 +1,448 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.google.common.collect.Ordering;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.RowN;
+import io.crate.jobs.ExecutionState;
+import io.crate.operation.RowUpstream;
+import io.crate.operation.projectors.sorting.OrderingByPosition;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class BlockingSortingQueuedRowDownstream implements RowMerger {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(BlockingSortingQueuedRowDownstream.class);
+
+    /**
+     * To prevent the Handles to block and unblock frequently when the queue size is reached
+     * we use two limits: MAX_QUEUE_SIZE and RESUME_AFTER.
+     *
+     * The handle is blocked when the queue has reached MAX_QUEUE_SIZE
+     * and unblocked when the queue has RESUME_AFTER or less rows
+     */
+    public static int MAX_QUEUE_SIZE = 1000;
+    public static int RESUME_AFTER = 10;
+
+    private final Ordering<Object[]> ordering;
+    private final List<BlockingSortingQueuedRowDownstreamHandle> downstreamHandles = new ArrayList<>();
+    private final List<RowUpstream> upstreams = new ArrayList<>();
+    private final AtomicInteger remainingUpstreams = new AtomicInteger(0);
+    private final AtomicBoolean downstreamAborted = new AtomicBoolean(false);
+    private final RowReceiver downstreamRowReceiver;
+    private final SharedArrayRef lowestToEmit;
+    private final int rowSize;
+    private final AtomicInteger runningHandles = new AtomicInteger(0);
+
+    private int pauseCount = 0;
```

Comment:
afaik this is only used for trace outputs, maybe remove or comment it. also, to get accurate values it should be atomic because its increment by multiple threads